ramips: rename interfaces for tplink er605v2

mar-kolya opened the pull request:
Currently eth1 (which is the first "lan" interface) doesn't work on this device. During boot the following can be seen in logs:

```
[    2.252804] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    2.266060] mtk_soc_eth 1e100000.ethernet eth0: mediatek frame engine at 0xbe100000, irq 19
[    2.277889] mtk_soc_eth 1e100000.ethernet eth1: mediatek frame engine at 0xbe100000, irq 19
...
[    2.355157] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    2.390312] mt7530-mdio mdio-bus:1f: configuring for fixed/rgmii link mode
[    2.398597] mt7530-mdio mdio-bus:1f: Link is Up - 1Gbps/Full - flow control rx/tx
[    2.403872] mt7530-mdio mdio-bus:1f eth1 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7530 PHY] (irq=21)
[    2.416988] mtk_soc_eth 1e100000.ethernet eth0: error -17 registering interface eth1
[    2.426973] mt7530-mdio mdio-bus:1f eth2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7530 PHY] (irq=22)
[    2.440996] mt7530-mdio mdio-bus:1f eth3 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7530 PHY] (irq=23)
[    2.454405] mt7530-mdio mdio-bus:1f eth4 (uninitialized): PHY [mt7530-0:04] driver [MediaTek MT7530 PHY] (irq=24)
[    2.467198] mtk_soc_eth 1e100000.ethernet eth0: entered promiscuous mode
[    2.474117] DSA: tree 0 setup
...
[    6.820998] mtk_soc_eth 1e100000.ethernet dsa: renamed from eth0
[    6.919904] mtk_soc_eth 1e100000.ethernet wan: renamed from eth1
```

So the problem seems to be the fact that built-in gmacs get default names (eth0/eth1) and are renamed after switch ports are initialized. This means that when switch port with name `eth1` is brought up this name is still used by gmac1 causing switch port's init to fail.

This patch just renames the ports to avoid name collision. Note: this will break existing configs for this device because it renames all the ports. This should not be major problem because this device doesn't have a proper OEM image and is only flashable with serial access, meaning there should not be many users.

musashino205 commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -140,7 +140,7 @@ ramips_setup_interfaces()
 		;;
 	tplink,er605-v2)
-		ucidef_set_interfaces_lan_wan "eth1 eth2 eth3 eth4" "eth0"
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" "wan"
```
This is the same as the default (` *) `) and you can drop this configuration to use the default.

mar-kolya replied:
thanks! Updated

GoetzGoerisch replied:
As the device is labeled port 1-5, where port 1 is WAN, and the others can are labeled WAN/LAN 2-3 and LAN 4-5.

I would propose that we rename the LAN ports to LAN2-5 to make it consistent with the device labeling?

See: https://static.tp-link.com/upload/image-line/ER605_V2_VPN_Load_Balance_Router_01_large_20211125084350g.jpg

mar-kolya replied:
TBH, I do not have a strong opinion about this, I can update the PR. I just would like to have all ports working :).
Would you prefer to also have `wan1`?

GoetzGoerisch replied:
Yes this is a great idea. 
`wan1` and `lan2-5` then for the ports would fit perfectly.

mar-kolya committed:
ramips: rename interfaces for tplink er605v2

Currently eth1 (which is the first "lan" interface) doesn't work on this device.
During boot the following can be seen in logs:

```
[    2.252804] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    2.266060] mtk_soc_eth 1e100000.ethernet eth0: mediatek frame engine at 0xbe100000, irq 19
[    2.277889] mtk_soc_eth 1e100000.ethernet eth1: mediatek frame engine at 0xbe100000, irq 19
...
[    2.355157] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    2.390312] mt7530-mdio mdio-bus:1f: configuring for fixed/rgmii link mode
[    2.398597] mt7530-mdio mdio-bus:1f: Link is Up - 1Gbps/Full - flow control rx/tx
[    2.403872] mt7530-mdio mdio-bus:1f eth1 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7530 PHY] (irq=21)
[    2.416988] mtk_soc_eth 1e100000.ethernet eth0: error -17 registering interface eth1
[    2.426973] mt7530-mdio mdio-bus:1f eth2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7530 PHY] (irq=22)
[    2.440996] mt7530-mdio mdio-bus:1f eth3 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7530 PHY] (irq=23)
[    2.454405] mt7530-mdio mdio-bus:1f eth4 (uninitialized): PHY [mt7530-0:04] driver [MediaTek MT7530 PHY] (irq=24)
[    2.467198] mtk_soc_eth 1e100000.ethernet eth0: entered promiscuous mode
[    2.474117] DSA: tree 0 setup
...
[    6.820998] mtk_soc_eth 1e100000.ethernet dsa: renamed from eth0
[    6.919904] mtk_soc_eth 1e100000.ethernet wan: renamed from eth1
```

So the problem seems to be the fact that built-in gmacs get default
names (eth0/eth1) and are renamed after switch ports are initialized. This means
that when switch port with name `eth1` is brought up this name is still used by
gmac1 causing switch port's init to fail.

This patch just renames the ports to avoid name collision.
Note: this will break existing configs for this device because it renames all
the ports. This should not be major problem because this device doesn't have a
proper OEM image and is only flashable with serial access, meaning there should
not be many users.

Signed-off-by: Nikolay Martynov <mar.kolya@gmail.com>

GoetzGoerisch commented:
> This should not be major problem because this device doesn't have a proper OEM image and is only flashable with serial access, meaning there should not be many users.

I need to disagree to this, flashing is easily be done following the procedure describe in the device page.
Serial access is not required.


Furthermore in 5.15.161 I do not see these changes:
```
[    2.812970] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    2.827858] mtk_soc_eth 1e100000.ethernet dsa: mediatek frame engine at 0xbe100000, irq 21
[    2.840258] mtk_soc_eth 1e100000.ethernet eth0: mediatek frame engine at 0xbe100000, irq 21
[    2.850037] i2c_dev: i2c /dev entries driver
[    2.858109] NET: Registered PF_INET6 protocol family
[    2.865952] Segment Routing with IPv6
[    2.869700] In-situ OAM (IOAM) with IPv6
[    2.873821] NET: Registered PF_PACKET protocol family
[    2.879010] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    2.892577] 8021q: 802.1Q VLAN Support v1.8
[    2.904235] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    2.935205] mt7530-mdio mdio-bus:1f: configuring for fixed/rgmii link mode
[    2.942933] mt7530-mdio mdio-bus:1f: Link is Up - 1Gbps/Full - flow control rx/tx
[    2.944647] mt7530-mdio mdio-bus:1f eth1 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7530 PHY] (irq=23)
[    2.963779] mt7530-mdio mdio-bus:1f eth2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7530 PHY] (irq=24)
[    2.976677] mt7530-mdio mdio-bus:1f eth3 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7530 PHY] (irq=25)
[    2.989749] mt7530-mdio mdio-bus:1f eth4 (uninitialized): PHY [mt7530-0:04] driver [MediaTek MT7530 PHY] (irq=26)
[    3.002170] DSA: tree 0 setup
...
[    9.670831] mtk_soc_eth 1e100000.ethernet dsa: configuring for fixed/rgmii link mode
[    9.683705] mtk_soc_eth 1e100000.ethernet dsa: Link is Up - 1Gbps/Full - flow control rx/tx
[    9.692420] mt7530-mdio mdio-bus:1f eth1: configuring for phy/gmii link mode
...
[   31.400200] mtk_soc_eth 1e100000.ethernet dsa: Link is Down
[   31.428867] mtk_soc_eth 1e100000.ethernet dsa: configuring for fixed/rgmii link mode
[   31.437195] mtk_soc_eth 1e100000.ethernet dsa: Link is Up - 1Gbps/Full - flow control rx/tx
[   31.442881] device dsa entered promiscuous mode
[   31.450697] mt7530-mdio mdio-bus:1f eth1: configuring for phy/gmii link mode
[   31.458694] IPv6: ADDRCONF(NETDEV_CHANGE): dsa: link becomes ready
[   31.466932] br-lan: port 1(eth1) entered blocking state
[   31.472242] br-lan: port 1(eth1) entered disabled state
[   31.480137] device eth1 entered promiscuous mode
[   31.683255] mt7530-mdio mdio-bus:1f eth2: configuring for phy/gmii link mode
[   31.693219] br-lan: port 2(eth2) entered blocking state
[   31.698720] br-lan: port 2(eth2) entered disabled state
[   31.708556] device eth2 entered promiscuous mode
[   31.735852] mt7530-mdio mdio-bus:1f eth3: configuring for phy/gmii link mode
[   31.747165] br-lan: port 3(eth3) entered blocking state
[   31.752482] br-lan: port 3(eth3) entered disabled state
[   31.761311] device eth3 entered promiscuous mode
[   31.798145] mt7530-mdio mdio-bus:1f eth4: configuring for phy/gmii link mode
[   31.807716] br-lan: port 4(eth4) entered blocking state
[   31.813016] br-lan: port 4(eth4) entered disabled state
[   31.821759] device eth4 entered promiscuous mode
[   31.942527] mtk_soc_eth 1e100000.ethernet eth0: PHY [mdio-bus:00] driver [MediaTek MT7530 PHY] (irq=POLL)
[   31.952346] mtk_soc_eth 1e100000.ethernet eth0: configuring for phy/rgmii link mode
[   33.592804] mt7530-mdio mdio-bus:1f eth4: Link is Up - 10Mbps/Full - flow control rx/tx
[   33.601094] br-lan: port 4(eth4) entered blocking state
[   33.606480] br-lan: port 4(eth4) entered forwarding state
...
[   34.286540] mt7530-mdio mdio-bus:1f eth3: Link is Up - 1Gbps/Full - flow control off
[   34.294634] br-lan: port 3(eth3) entered blocking state
[   34.299961] br-lan: port 3(eth3) entered forwarding state
[   34.772541] mt7530-mdio mdio-bus:1f eth2: Link is Up - 1Gbps/Full - flow control rx/tx
[   34.780803] br-lan: port 2(eth2) entered blocking state
[   34.786186] br-lan: port 2(eth2) entered forwarding state
[   35.153680] mtk_soc_eth 1e100000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
[   35.162225] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready

```

mar-kolya commented:
> Furthermore in 5.15.161 I do not see these changes:

This problem exists on current snapshot, which runs `6.6.35`. I suspect this may be related to 3ea6125c501e3adc81227cc69fb9e28eec69f2b8 but I didn't specifically verify that.

> I need to disagree to this, flashing is easily be done following the procedure describe in the device page. Serial access is not required.

My information is outdated then :). Unfortunately I do not know if there is a better way to handle such interface renames that provides better compatibility.

GoetzGoerisch commented:
> Unfortunately I do not know if there is a better way to handle such interface renames that provides better compatibility.

@arinc9 how to cope with that?

I'm asuming, that many people are running ER605v2 with 23.05 currently and this will break, as soon as they upgrade to 24.xx if I see this correct.

mar-kolya commented:
My apologies, looks like I pasted wrong output into the PR description: this is the one from early fixes I had, not one with the bug visible. I'll update this later today.

arinc9 commented:
> My apologies, looks like I pasted wrong output into the PR description: this is the one from early fixes I had, not one with the bug visible. I'll update this later today.

I was confused reading the kernel log as I thought that DSA detected the already existing eth1 and initialised port@1 with lan1 as the network interface name instead.

You can just rename port@1 to lan1 and adjust ucidef_set_interfaces_lan_wan accordingly, without touching anything else.

This change of behaviour is not related to my patch mentioned above.

mar-kolya commented:
> You can just rename port@1 to lan1 and adjust ucidef_set_interfaces_lan_wan accordingly, without touching anything else.

I'm not sure I fully understand this suggestion.
Do you mean I can make a patch that renames `eth1` to `lan1` in dts and leaves other interfaces unchanged? So there will be `eth0, lan1, eth2, ...`.
Or do you mean to rename interface later during the boot process the config somehow? 

If it's the former it would seem like strange naming. 
If it's the latter I do not think it will work - interface fails to create when names clash. Once again - sorry for the wrong copy-paste in the description. I need to rebuild firmware without the patch to get correct output with the bug to properly illustrate this.

arinc9 commented:
I don't need to see the kernel log with the issue. The label property is not supposed to represent the network interface name of the MAC on software. DSA does this anyway, which is tolerated, for now. I'd rename all switch ports to swp0-4 but there's the request that old configurations shall not be affected. So the remaining option is to only rename the already problematic port. Rename it to swp1 if you'd like.

GoetzGoerisch commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -139,9 +139,6 @@ ramips_setup_interfaces()
 		ucidef_set_interface_lan "lan0 lan1 lan2 lan3"
 		;;
-	tplink,er605-v2)
-		ucidef_set_interfaces_lan_wan "eth1 eth2 eth3 eth4" "eth0"
```
Would be then
`ucidef_set_interfaces_lan_wan "lan2 lan3 lan4 lan5" "wan"`

GoetzGoerisch commented on target/linux/ramips/dts/mt7621_tplink_er605-v2.dts:
```diff
@@ -84,7 +84,7 @@
 &gmac1 {
 	status = "okay";
-	label = "eth0";
+	label = "wan";
```
as discussed `wan` or `wan1` both would be ok from my PoV

mar-kolya replied:
All should be updated now. Thanks!

mar-kolya commented:
@GoetzGoerisch thanks for approving this PR! It looks like it requires 5 more approvals. Is there anything I can/should do to help move this PR forward? Thanks!

GoetzGoerisch commented:
Wait

hauke commented:
When someone upgrades the device the old network configuration will not work any more and the user has to get into failsafe mode to recover.
Please increase the `DEVICE_COMPAT_VERSION` to `2.0` to inform the user about this case and give them the option to not migrate the configuration.

mar-kolya commented:
> Please increase the `DEVICE_COMPAT_VERSION` to `2.0` to inform the user about this case and give them the option to not migrate the configuration.

Updated now. Thanks!

DragonBluep commented on target/linux/ramips/dts/mt7621_tplink_er605-v2.dts:
```diff
@@ -84,7 +84,7 @@
 &gmac1 {
 	status = "okay";
-	label = "eth0";
+	label = "wan1";
```
If there is only one WAN port, `label = "wan";` should be enough.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1,4 +1,4 @@
-#
+
```
Do not touch it.

mar-kolya replied:
Sorry, fixed

DragonBluep commented on target/linux/ramips/dts/mt7621_tplink_er605-v2.dts:
```diff
@@ -97,22 +97,22 @@
 		port@1 {
 			status = "okay";
-			label = "eth1";
+			label = "lan2";
```
Why start from `2`?

mar-kolya replied:
This was originally suggested by @GoetzGoerisch in this comment: https://github.com/openwrt/openwrt/pull/15865#discussion_r1666382125. There is also picture of the device which are number from 1 (wan, hence wan1) to 2 (lan2) an d so on to 5 (lan5) - so there is reasoning behind this.

GoetzGoerisch replied:
On the device the ports are numbered 1 to 5.

- Port 1 is WAN
- Port 2 is WAN/LAN
- Port 3 is WAN/LAN
- Port 4 is LAN
- Port 5 is LAN

Therefore, I suggested the above labeling.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2551,6 +2551,8 @@ TARGET_DEVICES += tplink_ec330-g5u-v1
 define Device/tplink_er605-v2
   $(Device/nand)
+  DEVICE_COMPAT_VERSION := 2.0
+  DEVICE_COMPAT_MESSAGE := Config cannot be migrated because interface names have changed
```
You also need to increase version number in:
https://github.com/openwrt/openwrt/blob/06b37a5856ac7d0a2ddc2c0745ac1da3a01688d6/target/linux/ramips/mt7621/base-files/etc/board.d/05_compat-version#L11

mar-kolya replied:
Updated, thanks!

GoetzGoerisch commented:
@hauke @musashino205 @DragonBluep anything missing from your side?

GoetzGoerisch commented:
@mar-kolya do you run this image on your device?

Do one really have to force an upgrade here? I understood, that the configuration cannot be carried over, but thought that the image is valid for sysupgrade.
Or does this relate to #11739

mar-kolya commented:
> @mar-kolya do you run this image on your device?

Yes, I run ~latest master with this applied on my device, things seem to be greate.

> 
> Do one really have to force an upgrade here? I understood, that the configuration cannot be carried over, but thought that the image is valid for sysupgrade.

I'm not sure I fully understand what the question is here. If the question is 'can one update device using sysupgrade image (instead of 'factory'), then I think the answer is 'yes'. At least that's how I did this on my device. :)

> Or does this relate to #11739

I think this is completely untelated.

GoetzGoerisch commented:
Thank you @mar-kolya.

I'm seeing an `image check failed` when I upload a self-compiled image from `main` with this commit cherry-picked. 
https://github.com/GoetzGoerisch/openwrt/tree/feat/er605

One could only FORCE flash this image.

Djfe commented:
I already suggested making it a minor update (1.1 instead of 2.0) so you can run `sysupgrade -n URL/PATH` still.
I don't see how renaming interfaces is supposed to be a major update that requires a factory install.

mar-kolya commented:
> Thank you @mar-kolya.
> 
> I'm seeing an `image check failed` when I upload a self-compiled image from `main` with this commit cherry-picked. https://github.com/GoetzGoerisch/openwrt/tree/feat/er605
> 
> One could only FORCE flash this image.

Well, yes, it asks for force. I sort of assume that is expected, sorry about that.

mar-kolya commented:
> I already suggested making it a minor update (1.1 instead of 2.0) so you can run `sysupgrade -n URL/PATH` still. I don't see how renaming interfaces is supposed to be a major update that requires a factory install.

I think @hauke above suggested making it 2.0 (https://github.com/openwrt/openwrt/pull/15865#issuecomment-2241731881).

I think rationale behind this is that renaming interfaces may potentially break all their configuration resulting in device being non-accessible. I'm not entirely sure what the policy is about this.

I think this device is already `1.1`, should I make it `1.2`?

Thanks!

Djfe commented:
I see, yes 1.2 would work.
I would've also trusted Hauke.
But switching to DSA was also a minor increase and it also caused changes to the interface config 
https://github.com/openwrt/openwrt/commit/1eac573b530425c80c046eaa0b373b5e63d92451

mar-kolya commented:
> I see, yes 1.2 would work. I would've also trusted Hauke. But switching to DSA was also a minor increase and it also caused changes to the interface config [1eac573](https://github.com/openwrt/openwrt/commit/1eac573b530425c80c046eaa0b373b5e63d92451)

GoetzGoerisch commented:
From my experience and testing.
Coming from 23.05.5 upgrading two SNAPSHOT the device is broken anyway as the `eth1` is dead. This bite me on the first attempt, because this was my management port.

Therefore I would appreciate a merge.

The devic config has issues with the LED config as well, which is broken in respect to 23.05.

@mar-kolya Thank you for your effort.

Djfe commented:
which led's? usb, system, power or link/activity?

GoetzGoerisch commented:
There are naming collisions and the system led gets a different name and my 23.05 configuration breaks.

Additionally the USB LED is permanent on due to this in the default config.
I will write a proper issue in the next days after I reproduced it.

Djfe commented:
wait
That happens on a fresh install? (`sysupgrade -n`)

GoetzGoerisch commented:
Yes, this was my initial observertion, but I would like to review and reproduce it.
